File: Languages/C/English/Logic/Logic.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

// Function to display the greeting
void greeting() {
    printf("Hello, welcome!\n");
}

// Sum function
int sum(int a, int b) {
    return a + b;
}

// String manipulation function
void stringManipulation() {
    char name[] = "João";
    char greetingMessage[50];
    sprintf(greetingMessage, "Hello, %s!", name);  // Concatenating strings
    printf("%s\n", greetingMessage);  // Displays: Hello, João!

    char word[] = "programming";
    int length = strlen(word);  // Checks the length of the string
    int contains = strstr(word, "gram") != NULL;  // Checks if the word contains "gram"
    printf("Word length: %d\n", length);
    printf("Contains 'gram': %d\n", contains);
}

// Array example function
void arrayExample() {
    int numbers[] = {1, 2, 3, 4, 5};  // Declaration and initialization of an array
    printf("First number: %d\n", numbers[0]);  // Accesses the first value
    printf("Last number: %d\n", numbers[sizeof(numbers)/sizeof(numbers[0]) - 1]);  // Accesses the last value
}

// Input and output function
void inputOutput() {
    char name[50];
    printf("Enter your name: ");
    fgets(name, sizeof(name), stdin);
    name[strcspn(name, "\n")] = '\0';  // Removes the newline from the string
    printf("Hello, %s\n", name);
}

// Class and object example function (using structs)
typedef struct {
    char name[50];
    int age;
} Person;

// Presentation method
void introduce(Person *person) {
    printf("Hello, my name is %s and I am %d years old.\n", person->name, person->age);
}

// Example of 'this' function (using pointers)
void exampleThis() {
    typedef struct {
        char model[50];
    } Car;

    Car car = {"Beetle"};
    printf("Car model: %s\n", car.model);
}

// Inheritance function (simulated with structs and pointers)
typedef struct {
    char name[50];
} Animal;

void makeAnimalSound(Animal *animal) {
    printf("The animal makes a sound.\n");
}

typedef struct {
    Animal base;  // Inheriting from Animal
} Dog;

void makeDogSound(Dog *dog) {
    printf("The dog barks.\n");
}

// Encapsulation example function (simulated with functions)
typedef struct {
    int balance;
} BankAccount;

void deposit(BankAccount *account, int amount) {
    if (amount > 0) {
        account->balance += amount;
    }
}

int getBalance(BankAccount *account) {
    return account->balance;
}

// Exception example function (simulated with an if for division by zero)
void exceptionExample() {
    int result;
    if (0 != 0) {
        printf("Error: Division by zero\n");
    } else {
        result = 10 / 0;  // Division by zero
    }
}

int main() {
    // 1. Variables and Data Types
    int age = 25;  // Integer type
    float height = 1.75;  // Float type
    char name[] = "João";  // String type
    bool student = true;  // Boolean type

    printf("Age: %d\n", age);
    printf("Height: %.2f\n", height);
    printf("Name: %s\n", name);
    printf("Student: %s\n", student ? "True" : "False");

    // 2. Operators
    int sumResult = 10 + 5;  // Addition
    int remainder = 10 % 3;  // Modulo
    printf("Sum: %d\n", sumResult);
    printf("Remainder: %d\n", remainder);

    bool comparisonResult = 10 > 5;  // Greater than
    bool equality = 5 == 5;  // Equality
    printf("10 > 5: %s\n", comparisonResult ? "True" : "False");
    printf("5 == 5: %s\n", equality ? "True" : "False");

    bool andCondition = (10 > 5) && (5 < 10);  // Logical 'AND'
    bool orCondition = (10 > 5) || (5 > 10); // Logical 'OR'
    printf("AND Condition: %s\n", andCondition ? "True" : "False");
    printf("OR Condition: %s\n", orCondition ? "True" : "False");

    // 3. Conditional Structures
    int personAge = 20;
    if (personAge >= 18) {
        printf("You are an adult.\n");
    } else {
        printf("You are a minor.\n");
    }

    // 4. Loops
    for (int i = 1; i <= 5; i++) {
        printf("%d\n", i);
    }

    int j = 1;
    while (j <= 5) {
        printf("%d\n", j);
        j++;
    }

    int k = 1;
    do {
        printf("%d\n", k);
        k++;
    } while (k <= 5);

    // 5. Functions
    greeting();
    int sumResultFunction = sum(5, 3);
    printf("Sum result: %d\n", sumResultFunction);

    // 6. Input and Output
    inputOutput();

    // 7. String Manipulation
    stringManipulation();

    // 8. Arrays
    arrayExample();

    // 9. Class and Objects
    Person person1 = {"Maria", 30};
    introduce(&person1);

    // 10. Concept of `this`
    exampleThis();

    // 11. Inheritance
    Dog dog = {{"Rex"}};
    makeDogSound(&dog);

    // 12. Encapsulation
    BankAccount account = {0};
    deposit(&account, 500);
    printf("Account balance: %d\n", getBalance(&account));

    // 13. Exceptions
    exceptionExample();

    return 0;
}

File: Languages/Dart/English/Logic/Logic.dart
// 1. Variables and Data Types
// Variables are like "boxes" where we store information. Each box has a specific type.

void main() {
  int age = 25; // Integer type: stores whole numbers
  double height = 1.75; // Floating-point type: stores numbers with decimal places
  String name = "John"; // String type: stores a sequence of characters
  bool student = true; // Boolean type: stores true or false

  print("Age: $age");
  print("Height: $height");
  print("Name: $name");
  print("Student: $student");

  // 2. Operators
  // Arithmetic, comparison, and logical operators.

  // Arithmetic Operators
  int sum = 10 + 5; // Addition
  int remainder = 10 % 3; // Modulus (remainder of division)
  print("Sum: $sum");
  print("Remainder: $remainder");

  // Comparison Operators
  bool result = 10 > 5; // Checks if 10 is greater than 5
  bool equals = 5 == 5; // Checks if 5 is equal to 5
  print("10 > 5: $result");
  print("5 == 5: $equals");

  // Logical Operators
  bool andCondition = (10 > 5) && (5 < 10); // Logical AND: both conditions need to be true
  bool orCondition = (10 > 5) || (5 > 10); // Logical OR: one condition needs to be true
  print("AND Condition: $andCondition");
  print("OR Condition: $orCondition");

  // 3. Conditional Structures
  // Conditions to make decisions in the program.

  int personAge = 20;
  if (personAge >= 18) {
    print("You are an adult.");
  } else {
    print("You are a minor.");
  }

  // Switch Case: an alternative to if-else when we have multiple possible conditions
  int number = 2;
  switch (number) {
    case 1:
      print("It's one");
      break;
    case 2:
      print("It's two");
      break;
    default:
      print("Other number");
  }

  // 4. Loops
  // Used when we want to repeat a block of code multiple times.

  for (int i = 1; i <= 5; i++) {
    print(i); // Prints 1, 2, 3, 4, 5
  }

  int j = 1;
  while (j <= 5) {
    print(j); // Prints 1, 2, 3, 4, 5
    j++; // Increment the value of j
  }

  int k = 1;
  do {
    print(k); // Prints 1, 2, 3, 4, 5
    k++;
  } while (k <= 5);

  // 5. Functions (Methods)
  // Functions are used to organize and reuse blocks of code.

  greeting();
  int sumResult = sumFunc(5, 3);
  print("Sum result: $sumResult");

  // 6. Input and Output
  // Used to interact with the user, capturing inputs and displaying outputs.
  // Dart does not have the prompt function, so other methods would be needed.

  // 7. String Manipulation
  stringManipulation();

  // 8. Arrays
  exampleArrays();

  // 9. Class and Objects
  exampleClassObjects();

  // 10. The concept of `this`
  exampleThis();

  // 11. Inheritance
  exampleInheritance();

  // 12. Encapsulation
  exampleEncapsulation();

  // 13. Exceptions
  exampleException();

  // 14. Working with `ArrayList`
  exampleArrayList();
}

// Functions
void greeting() {
  print("Hello, welcome!");
}

int sumFunc(int a, int b) {
  return a + b;
}

void stringManipulation() {
  String name = "John";
  String greetingMessage = "Hello, $name!"; // Concatenating strings
  print(greetingMessage); // Displays: Hello, John!

  String word = "programming";
  int length = word.length; // Checks the length of the string
  bool contains = word.contains("gram"); // Checks if the word contains "gram"
  print("Word length: $length");
  print("Contains 'gram': $contains");
}

void exampleArrays() {
  List<int> numbers = [1, 2, 3, 4, 5]; // Array declaration and initialization
  print("First number: ${numbers[0]}"); // Accesses the first value
  print("Last number: ${numbers[numbers.length - 1]}"); // Accesses the last value
}

class Person {
  String name;
  int age;

  Person(this.name, this.age);

  void introduce() {
    print("Hello, my name is $name and I am $age years old.");
  }
}

void exampleClassObjects() {
  Person person1 = Person("Nick", 30);
  person1.introduce();
}

void exampleThis() {
  class Car {
    String model;

    Car(this.model);

    void showModel() {
      print("Car model: $model");
    }
  }

  var car = Car("Beetle");
  car.showModel();
}

class Animal {
  String name;

  Animal(this.name);

  void makeSound() {
    print("The animal makes a sound.");
  }
}

class Dog extends Animal {
  Dog(String name) : super(name);

  @override
  void makeSound() {
    print("The dog barks.");
  }
}

void exampleInheritance() {
  Dog dog = Dog("Rex");
  dog.makeSound(); // The overridden method is called
}

class BankAccount {
  double _balance = 0; // Private attribute

  void deposit(double amount) {
    if (amount > 0) {
      _balance += amount;
    }
  }

  double getBalance() {
    return _balance; // Access to balance through a public method
  }
}

void exampleEncapsulation() {
  BankAccount account = BankAccount();
  account.deposit(500);
  print("Account balance: ${account.getBalance()}");
}

void exampleException() {
  try {
    int result = 10 ~/ 0; // Division by zero
  } catch (e) {
    print("Error: ${e}");
  }
}

void exampleArrayList() {
  List<String> list = ["Java", "Python", "C++"];

  print("List size: ${list.length}"); // Displays the list size
  print("First item: ${list[0]}"); // Accesses the first item of the list
}
